typescript/sdk/src/core/EvmIcaModule.ts

```diff
@@ -15,33 +29,182 @@ import {
+    );
+    this.domainId = multiProvider.getDomainId(args.chain);
+    this.chainId = multiProvider.getChainId(args.chain);
   }
 
-  public async read(): Promise<InterchainAccountConfig> {
-    throw new Error('Method not implemented.');
+  public async read(): Promise<DerivedIcaRouterConfig> {
+    return this.icaRouterReader.deriveConfig(
+      this.args.addresses.interchainAccountRouter,
+    );
   }
 
   public async update(
-    _config: InterchainAccountConfig,
+    expectedConfig: InterchainAccountConfig,
+  ): Promise<AnnotatedEV5Transaction[]> {
+    const actualConfig = await this.read();
+
+    const transactions: AnnotatedEV5Transaction[] = [
+      ...(await this.updateRemoteRoutersEnrollment(
+        actualConfig.remoteIcaRouters,
+        expectedConfig.remoteIcaRouters,
+      )),
+      ...proxyAdminUpdateTxs(
+        this.chainId,
+        this.args.addresses.interchainAccountIsm,
+        actualConfig,
+        expectedConfig,
+      ),
+    ];
+
+    return transactions;
+  }
+
+  private async updateRemoteRoutersEnrollment(
+    actualConfig: DerivedIcaRouterConfig['remoteIcaRouters'],
+    expectedConfig: InterchainAccountConfig['remoteIcaRouters'] = {},
+  ): Promise<AnnotatedEV5Transaction[]> {
+    const transactions: AnnotatedEV5Transaction[] = [
+      ...(await this.getEnrollRemoteIcaRoutersTxs(
+        actualConfig,
+        expectedConfig,
+      )),
+      ...(await this.getUnenrollRemoteIcaRoutersTxs(
+        actualConfig,
+        expectedConfig,
+      )),
+    ];
+
+    return transactions;
+  }
+
+  private async getEnrollRemoteIcaRoutersTxs(
+    actualConfig: Readonly<DerivedIcaRouterConfig['remoteIcaRouters']>,
+    expectedConfig: Readonly<InterchainAccountConfig['remoteIcaRouters']> = {},
+  ): Promise<AnnotatedEV5Transaction[]> {
+    const transactions: AnnotatedEV5Transaction[] = [];
+
+    const routesToEnroll = Array.from(
+      difference(
+        new Set(Object.keys(expectedConfig)),
+        new Set(Object.keys(actualConfig)),
+      ),
+    );
+
+    if (routesToEnroll.length === 0) {
+      return transactions;
+    }
+
+    const domainsToEnroll: string[] = [];
+    const remoteDomainIca: string[] = [];
+    const remoteIsm: string[] = [];
+
+    for (const domainId of routesToEnroll) {
+      domainsToEnroll.push(domainId);
+      remoteDomainIca.push(addressToBytes32(expectedConfig[domainId].address));
+      remoteIsm.push(
+        expectedConfig[domainId].interchainSecurityModule
+          ? addressToBytes32(expectedConfig[domainId].interchainSecurityModule!)
+          : ethers.utils.hexZeroPad('0x', 32),
+      );
+    }
+
+    const remoteTransactions: AnnotatedEV5Transaction[] = domainsToEnroll.map(
+      (domainId) => ({
+        annotation: `Enrolling InterchainAccountRouter on domain ${this.domainId} on InterchainAccountRouter at ${expectedConfig[domainId].address} on domain ${domainId}`,
+        chainId: Number(this.multiProvider.getChainId(domainId)),
```

Comment:
think this is no longer true due to https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/4798/files

Reply:
Good catch! Should probably use the new `getEvmChainId` method here